### PR TITLE
Unit test waitForNodes

### PR DESCRIPTION
**Pull request by asalkeld**
part of #382

### CODE REVIEW

File: pkg/upgrade/ready.go [hunk 1]
```diff
@@ -85,22 +85,30 @@ var daemonsetWhitelist = []struct {
...
 
-func WaitForNodes(ctx context.Context, cs *api.OpenShiftManagedCluster, kc *kubernetes.Clientset) error {
-	config := api.PluginConfig{AcceptLanguages: []string{"en-us"}}
-	authorizer, err := azureclient.NewAuthorizerFromContext(ctx)
-	if err != nil {
-		return err
+func (u *simpleUpgrader) waitForNodes(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
+	var err error
+	if u.kubeclient == nil {
+		u.kubeclient, err = managedcluster.ClientsetFromV1ConfigAndWait(ctx, cs.Config.AdminKubeconfig)
```
Comment (kargakis) [inline pkg/upgrade/ready.go]: Waiting for the api server here seems a bit redundant, we should have healthchecking code for the api server elsewhere already, no?

File: pkg/upgrade/ready_test.go [hunk 2]
```diff
@@ -238,3 +244,42 @@ func TestMasterIsReady(t *testing.T) {
...
+		{
+			name:       "nothing to wait for",
+			kubeclient: fake.NewSimpleClientset(),
+			cs: &api.OpenShiftManagedCluster{
+				Properties: &api.Properties{
+					AzProfile: &api.AzProfile{ResourceGroup: testRg},
+				},
+			},
+			wantErr: false,
+		},
```
Comment (kargakis) [inline pkg/upgrade/ready_test.go]: two more tests to add:
1. vm list fails
2. waiting for a ready node fails

File: pkg/upgrade/ready.go (hunk 1 above)
Comment (kargakis) [reply pkg/upgrade/ready.go]: This is probably fine.

File: pkg/upgrade/ready_test.go (hunk 2 above)
Comment (kargakis) [reply pkg/upgrade/ready_test.go]: waiting for a ready node fails is still missing from your test cases.

File: pkg/upgrade/ready_test.go [hunk 3]
```diff
@@ -238,3 +248,261 @@ func TestMasterIsReady(t *testing.T) {
...
+			virtualMachineScaleSetsClient := mock_azureclient.NewMockVirtualMachineScaleSetsClient(gmc)
+			virtualMachineScaleSetVMsClient := mock_azureclient.NewMockVirtualMachineScaleSetVMsClient(gmc)
+
+			mPage := mock_azureclient.NewMockVirtualMachineScaleSetVMListResultPage(gmc)
+			iPage := mock_azureclient.NewMockVirtualMachineScaleSetVMListResultPage(gmc)
+			cPage := mock_azureclient.NewMockVirtualMachineScaleSetVMListResultPage(gmc)
+
+			if len(tt.expect) > 0 {
+				mPage.EXPECT().Values().Return(tt.expect["master"])
+				mPage.EXPECT().Next()
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/upgrade/ready_test.go]: it's called here: https://github.com/openshift/openshift-azure/blob/master/pkg/upgrade/update.go#L126

File: pkg/upgrade/ready_test.go [hunk 4]
```diff
@@ -238,3 +248,261 @@ func TestMasterIsReady(t *testing.T) {
...
+				iPage.EXPECT().Values().Return(tt.expect["infra"])
+				iPage.EXPECT().Next()
+				cPage.EXPECT().Values().Return(tt.expect["compute"])
+				cPage.EXPECT().Next()
+			}
+			callTimes := func(vms []compute.VirtualMachineScaleSetVM) int {
+				if len(vms) > 0 {
+					return 2
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/upgrade/ready_test.go]: ok

File: pkg/upgrade/ready_test.go [hunk 5]
```diff
@@ -238,3 +248,261 @@ func TestMasterIsReady(t *testing.T) {
...
+	type args struct {
+		ctx context.Context
+		kc  *kubernetes.Clientset
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		// TODO: Add test cases.
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/upgrade/ready_test.go]: oops, this shouldn't be in this PR. I'll remove

File: pkg/upgrade/ready_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/upgrade/ready_test.go]: So this 2 is really `len(vms)+1`, is that right?

[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/upgrade/ready_test.go]: If so all this function can be collapsed to:
```go
callTimes := func(vms []compute.VirtualMachineScaleSetVM) int {
  return len(vms)+1
}
```

**Comment by asalkeld** [comment]
Only when you have a vm list of 0 or 1, if someone wants to add a another
vm to the list then that logic breaks.

**Comment by kargakis** [comment]
Is it that with every new vm you have one Next plus so then it should be
2*len(vms) in case we have vms?

**Comment by asalkeld** [comment]
No, it returns all in a list so you either have one call

No data
Or
Yes I have data, then no more data.

Angus

**Pull request merged**